Source/Engine/PackageBuilder/NormalizingPatternLinker.cs

```diff
@@ -0,0 +1,134 @@
+            Syntax result = base.VisitPattern(node);
+            fCurrentPattern = null;
+            fAllPatterns.Add(result);
+            return result;
+        }
+
+        protected internal override Syntax VisitInside(InsideSyntax node)
+        {
+            Syntax newInner = Visit(node.Inner);
+            Syntax newOuter = Visit(node.Outer);
+            if (!(newOuter is PatternReferenceSyntax))
+            {
+                var extractionFromFields = new List<Syntax>();
+                foreach (FieldSyntax field in fCurrentPattern.Fields)
+                    extractionFromFields.Add(Syntax.ExtractionFromField(field, field.Name));
+                var pattern = Syntax.Pattern(fCurrentPattern.Namespace, isSearchTarget: false,
+                    GetExtractedPatternName(), fCurrentPattern.Fields, newOuter);
+                newOuter = Syntax.PatternReference(pattern, extractionFromFields);
+                fExtractedPatterns.Add(pattern);
+            }
+            Syntax result = node.Update(newInner, newOuter);
+            return result;
+        }
+        
+        protected internal override Syntax VisitOutside(OutsideSyntax node)
+        {
+            Syntax newBody = Visit(node.Body);
+            Syntax newException = Visit(node.Exception);
+            if (!(newException is PatternReferenceSyntax))
+            {
+                var pattern = Syntax.Pattern(fCurrentPattern.Namespace, isSearchTarget: false,
+                    GetExtractedPatternName(), fCurrentPattern.Fields, newException);
+                newException = Syntax.PatternReference(pattern, Syntax.EmptyExtractionList());
+                fExtractedPatterns.Add(pattern);
+            }
+            Syntax result = node.Update(newBody, newException);
+            return result;
+        }
+        
+        protected internal override Syntax VisitHaving(HavingSyntax node)
+        {
+            Syntax newOuter = Visit(node.Outer);
+            Syntax newInner = Visit(node.Inner);
+            if (!(newInner is PatternReferenceSyntax))
+            {
+                var extractionFromFields = new List<Syntax>();
+                foreach (FieldSyntax field in fCurrentPattern.Fields)
+                    extractionFromFields.Add(Syntax.ExtractionFromField(field, field.Name));
+                var pattern = Syntax.Pattern(fCurrentPattern.Namespace, isSearchTarget: false,
+                    GetExtractedPatternName(), fCurrentPattern.Fields, newInner);
+                newInner = Syntax.PatternReference(pattern, extractionFromFields);
+                fExtractedPatterns.Add(pattern);
+            }
+            Syntax result = node.Update(newOuter, newInner);
+            return result;
+        }
+        
+        private string GetExtractedPatternName()
+        {
+            string result = $"{fCurrentPattern.Name}/{fNextExtractedPatternNumber}";
+            fNextExtractedPatternNumber++;
+            return result;
+        }
+    }
+}
```

Comment:
Cosmetics: please add line break here at the end of file, as I see we do it in other files.